application/common/models/Reset.php

```diff
@@ -43,14 +60,160 @@ public function rules()
     }
 
     /**
+     * @return array
+     */
+    public function fields()
+    {
+        return [
+            'uid',
+            'hasSupervisor' => $this->user->hasSupervisor(),
+            'hasSpouse' => $this->user->hasSpouse(),
+            'methods' => $this->user->getMaskedMethods(),
+        ];
+    }
+
+    /**
+     * @param User $user
+     * @param string [default=self::TYPE_PRIMARY] $type
+     * @param integer|null [default=null] $method_id
+     * @return Reset
+     * @throws NotFoundHttpException
+     * @throws \Exception
+     */
+    public static function findOrCreate($user, $type = self::TYPE_PRIMARY, $methodId = null)
+    {
+        /*
+         * Find existing or create new Reset
+         */
+        $reset = self::findOne(['user_id' => $user->id]);
+        if ($reset === null) {
+            $reset = new Reset();
+            $reset->user_id = $user->id;
+            /*
+             * Only set type/method if creating so that on subsequent restarts of process
+             * it does not reset method to primary
+             */
+            $reset->type = $type;
+            /*
+             * If $method_id is provided, make sure user owns it
+             */
+            if ($type == self::TYPE_METHOD && $methodId !== null) {
+                $method = Method::findOne(['user_id' => $user->id, 'id' => $methodId]);
+                if ( ! $method) {
+                    throw new NotFoundHttpException('Requested method not found', 1456608142);
+                }
+                $reset->method_id = $methodId;
+            }
+            /*
+             * Save new Reset
+             */
+            if ( ! $reset->save()) {
+                throw new \Exception('Unable to create new reset', 1456608028);
+            }
+        }
+
+        /*
+         * Send Reset to appropriate method
+         */
+        $reset->send();
```

Comment:
Should we be calling `send()` even if the Reset already existed? I would have guess from the name `findOrCreate(...)` that it would also automatically send the reset code.